src/lib/tls/tls12/msg_client_kex.cpp

```diff
@@ -133,6 +133,15 @@ Client_Key_Exchange::Client_Key_Exchange(Handshake_IO& io,
                                                           rng,
                                                           policy);
 
+         // RFC 8422 - 5.11.
+         //   With X25519 and X448, a receiving party MUST check whether the
+         //   computed premaster secret is the all-zero value and abort the
+         //   handshake if so, as described in Section 6 of [RFC7748].
+         if(curve_id == Group_Params::X25519 && CT::all_zeros(shared_secret.data(), shared_secret.size()).is_set())
+            {
+            throw TLS_Exception(Alert::InternalError, "Bad X25519 key exchange");
```

Comment:
Internal errors are meant for intrinsic errors of the implementation (e.g. out-of-memory). I guess:

```suggestion
            throw TLS_Exception(Alert::DecryptError, "Bad X25519 key exchange");
```

RFC 8446:
```
decrypt_error:  A handshake (not record layer) cryptographic
      operation failed, including being unable to correctly verify a
      signature or validate a Finished message or a PSK binder.
```

Reply:
I remember that I selected the same alert OpenSSL uses in the same situation. I was also a bit confused by the term "decrypt", because it is no decryption but a key exchange. However, I think you are right because it is indeed a "cryptographic operation". that has (somehow) "failed".